### DISCUSSION TITLE
expect() in jest fails multiple times for the same test

### DISCUSSION THREAD

**Discussion by covik** [pulumi/pulumi#9030] [Pulumi in Practice] [answered]
I have a following function I'd like to test:
```typescript
export function createNamespace(provider: k8s.Provider): k8s.core.v1.Namespace {
    return new k8s.core.v1.Namespace('primary-namespace', {
        metadata: {
            name: 'irrelevant-name'
        }
    }, {
        provider
    })
}
```

This is the test file:
```typescript
import * as pulumi from '@pulumi/pulumi'
import * as k8s from '@pulumi/kubernetes'

pulumi.runtime.setMocks({
    newResource: function(args: pulumi.runtime.MockResourceArgs): {id: string, state: any} {
        return {
            id: args.inputs.name + "_id",
            state: args.inputs
        }
    },
    call: function(args: pulumi.runtime.MockCallArgs) {
        return args.inputs
    }
})

describe('Infrastructure', function() {
    let manifests: typeof import('../../pulumi/create-kubernetes-manifests')

    beforeAll(async () => {
        manifests = await import('../../pulumi/create-kubernetes-manifests')
    })

    it('namespace name', () => {
        const provider: k8s.Provider = new k8s.Provider('fake-provider', {
            kubeconfig: ''
        })

        manifests.createNamespace(provider).metadata.name.apply(namespace => {
            expect(namespace).toBe('failure')
        })
    })
})
```

As you can see from the test body, expectation is that namespace name is 'failure' while real name is 'irrelevant-name' which makes the test fail.
That's expected behavior but the test fails four times:
```
node:internal/process/promises:246
          triggerUncaughtException(err, true /* fromPromise */);
          ^

[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "Error: expect(received).toBe(expected) // Object.is equality

Expected: "failure"
Received: "irrelevant-name"".] {
  code: 'ERR_UNHANDLED_REJECTION'
}

Node.js v17.3.0
node:internal/process/promises:246
          triggerUncaughtException(err, true /* fromPromise */);
          ^

[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "Error: expect(received).toBe(expected) // Object.is equality

Expected: "failure"
Received: "irrelevant-name"".] {
  code: 'ERR_UNHANDLED_REJECTION'
}

Node.js v17.3.0
node:internal/process/promises:246
          triggerUncaughtException(err, true /* fromPromise */);
          ^

[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "Error: expect(received).toBe(expected) // Object.is equality

Expected: "failure"
Received: "irrelevant-name"".] {
  code: 'ERR_UNHANDLED_REJECTION'
}

Node.js v17.3.0
node:internal/process/promises:246
          triggerUncaughtException(err, true /* fromPromise */);
          ^

[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "Error: expect(received).toBe(expected) // Object.is equality

Expected: "failure"
Received: "irrelevant-name"".] {
  code: 'ERR_UNHANDLED_REJECTION'
}

Node.js v17.3.0
 FAIL  src/try-pulumi/__tests__/pulumi.unit.ts
  ● Test suite failed to run

    Jest worker encountered 4 child process exceptions, exceeding retry limit

      at ChildProcessWorker.initialize (node_modules/jest-worker/build/workers/ChildProcessWorker.js:185:21)

```

This makes failing test take much more time to fail than it's needed. Jest says the exception is thrown from promise but I see no promises in the test. Is `apply()` doing something with the Promises behind the scenes or I'm missing something else here?

Jest v27.4.7
Node.js v17.3.0
Pulumi v3.21.0

**Comment by cnunciato** [comment, marked as answer]
I get the same error when I construct a test in this way (using slightly different resources), although I only get it once.

Jest [does handle promises](https://jestjs.io/docs/asynchronous), though, so I'm curious what you see when you explicitly convert the `pulumi.Output` to a promise and return that instead (or use some other approach, like passing a callback function to `it`). I've used this approach, for example:

In `index.ts`:

```typescript
import * as aws from "@pulumi/aws";

export const bucket = new aws.s3.Bucket("my-bucket", {
    // tags: { SomeTag: "some-value" }, <-- commented so as to fail
});
```

In `index.spec.ts`:

```typescript
import * as pulumi from "@pulumi/pulumi";
import "jest";
// ...

function promiseOf<T>(output: pulumi.Output<T>): Promise<T> {
    return new Promise(resolve => output.apply(resolve));
}

describe("Infrastructure", () => {
    let infra: typeof import("./index");

    beforeEach(async () => {
        infra = await import("./index");
    });

    describe("#bucket", () => {
        it("has tags", () => {
            return promiseOf(infra.bucket.tags).then(tags => {
                expect(tags).toBeDefined();
            });
        });
    });
});
```

Maybe this approach will work for you as well?

**Comment by covik** [comment]
I tried your approach in my codebase and I can confirm it fails once and exits immediately. I'm still curious why my initial approach does not function in the same way. Do you have some insights or clues?

**Reply by cnunciato** [reply]
Excellent, glad that worked. That'll take a bit of digging, but it's an interesting question -- will post back with what I can find.

**Reply by covik** [reply]
Hope the following information can help:
when I run the test from my initial question with `jest` in single process (`--runInBand`) the test fails only once.
Seems like jest has some kind of retry mechanism when dealing with promises in parallel execution which makes it fail multiple times.

**Reply by cnunciato** [reply]
My best explanation is that this happens because the test throws within the function that's passed to `apply()`, and that function is itself invoked within a promise (later, within the Pulumi SDK). If you follow the Node.js SDK implementation of `apply()`, you can trace things through: 

Here's where the `func` is passed into `apply`:

https://github.com/pulumi/pulumi/blob/b0bc4e3fdc4b1490187b05394788900939b68a89/sdk/nodejs/output.ts#L298-L302

And here's where it's invoked:

https://github.com/pulumi/pulumi/blob/b0bc4e3fdc4b1490187b05394788900939b68a89/sdk/nodejs/output.ts#L383

Alternatively, when you wrap the `pulumi.Output` in a promise (as in my example), and then return that promise to Jest, the exception is caught properly (by Jest), so you don't get the unhandled rejection error from Node.

**Reply by covik** [reply]
Thanks for the explanation!
Should this situation be noted in the Documentation assuming many people might be using jest?

**Reply by cnunciato** [reply]
Yes, I think so! I was also thinking it might make for a good blog post.

**Reply by cnunciato** [reply]
Incidentally @Covik if my first response did work out for you, you should be able to mark it as the answer. I'll post back here when we have more docs on this as well!

**Reply by covik** [reply]
True that, done!

> I'll post back here when we have more docs on this as well!

Sure, take your time, no emergencies here :smile:

**Comment by cnunciato** [comment]
Just stopping by to share that we wrote a post about working with Jest, in case anyone finds it helpful. Thank you again @Covik for asking this question! https://www.pulumi.com/blog/testing-pulumi-programs-with-jest/
